### ISSUE TITLE
invariants mining without label file

### ISSUE THREAD

**Issue by Rufaida94** [logpai/loglizer#33]
my problem is when I try to re-implement the InvariantsMiner_demo.py to work without the label file ( I don't care about measuring the correctness at this point) it produces a different set of invariants compared to the original InvariantsMiner_demo.py.

With my implementation I get the following results:
Invariant space dimension: 17
Mined 17 invariants: {(19,): [1], (0, 3): [-1.0, 1.0], (0, 4): [1.0, -1.0], (0, 5): [-17.0, 1.0], (0, 6): [-17.0, 1.0], (0, 10): [-17.0, 1.0], (0, 15): [-35.0, 1.0], (0, 16): [-9.0, 1.0], (0, 17): [1.0, -2.0], (1, 9): [1.0, -3.0], (1, 11): [-1.0, 1.0], (1, 14): [-1.0, 1.0], (1, 18): [1.0, -1.0], (5, 8): [1.0, -1.0], (5, 13): [-101.0, 1.0], (6, 8): [1.0, -1.0], (6, 13): [-101.0, 1.0]}

 
With the original InvariantsMiner_demo.py code I get the following results:

Invariant space dimension: 12
Mined 11 invariants: {(14,): [1], (0, 2): [1.0, -1.0], (0, 12): [-15.0, 1.0], (1, 6): [1.0, -3.0], (1, 8): [-1.0, 1.0], (1, 10): [-1.0, 1.0], (1, 13): [1.0, -1.0], (3, 4): [1.0, -1.0], (3, 7): [1.0, -1.0], (3, 9): [-101.0, 1.0], (3, 11): [-2.0, 1.0]}

can you please point out what I am doing wrong?

**Comment by Rufaida94** [comment]
here is my code:




import sys
sys.path.append('../')
from loglizer.models import InvariantsMiner
from loglizer import dataloader, preprocessing

struct_log = '../data/HDFS/HDFS_100k.log_structured.csv' # The structured log file
epsilon = 0.5 # threshold for estimating invariant space

if __name__ == '__main__':
    (x_train, _), (_, _) = dataloader.load_HDFS(struct_log,window='session',train_ratio=0.5)
    feature_extractor = preprocessing.FeatureExtractor()
    x_train = feature_extractor.fit_transform(x_train)
    print('Train phase:')
    model = InvariantsMiner(epsilon=epsilon)
    model.fit(x_train)
    y_train = model.predict(x_train)

    print('Test phase:')
   # Load another new log file. Here we use struct_log for demo only
    (x_test, _), (_, _) = dataloader.load_HDFS(struct_log, window='session')
    # Go through the same feature extraction process with training, using transform() instead
    x_test = feature_extractor.transform(x_test)
    # Finall make predictions and alter on anomaly cases
    y_test = model.predict(x_test)

**Comment by jimzhu** [comment]
Thanks for raising the issue. The problem is in the load_HDFS method. If you would like to compare with InvariantMiner_demo.py, you need to set train_ratio = 1. Since the train_ratio is only effect when label_file is given. We will take this into consideration later.

**Comment by Rufaida94** [comment]
do you mean just changing it to
  (x_train, _), (_, _) = dataloader.load_HDFS(struct_log,window='session',train_ratio=1)
?
because I tried that and it still produces the same output (17 mined invariants instead of 13)

**Comment by Rufaida94** [comment]
it will be great if you can revise the code and upload a version for invariants mining without the label file just like you did for the PCA. 

thanks

**Comment by jimzhu** [comment]
The demos have been updated for considering the issue you reported. Please note that we added a parameter to control how to split training and testing data. split_type=`uniform` or `sequential`.
